### PR TITLE
Expose more configurable PyTorch Lightning parameters

### LINKED ISSUE #189: Using `normalize` arguments in chemprop models, lots of sharp edges, training and inference possibly broken
Quite a few sharp edges around normalizing data, further investigation required.

### LINKED ISSUE #222: Unify LightningTrainer and GATv2LightningTrainer
We currently have a GATv2-specific PyTorch Lightning trainer. Should be able to streamline into one.

### LINKED ISSUE #228: Expand PyTorch Lightning configurability
Currently many trainer settings will simply assume default. No ability to configure optimizer or scheduler. Can expose more settings.

### PR DESCRIPTION
**Pull request by smcolby**
## Description
Currently many trainer settings will simply assume default. No ability to configure optimizer or scheduler. Can expose more settings.

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [X] Expose more trainer settings
  - [X] Expose scheduler settings
  - [X] Expose optimizer settings
  - [x] Unify GAT and default lightning trainers (@yxnyu)
  - [x] Test

## Status
- [x] Ready to go

Resolves #228 
Resolves #222 
May resolve #189

### CODE REVIEW

**Comment by hmacdope** [comment]
Currently GAT losses coming out as 0.0, looking into it.

**Comment by hmacdope** [comment]
> Currently GAT losses coming out as `np.NaN`, looking into it.

This was due to `NaN` inputs, something we need to handle better.

**Review by hmacdope — APPROVED** [review]
I think this is working now! Happy to move forward here and fix things as we go forward.

File: openadmet/models/architecture/gat.py [hunk 1]
```diff
@@ -232,40 +347,42 @@ def build(self, scaler=None, **kwargs):
...
         else:
             logger.warning("Model already exists, skipping build")
 
+
+    def make_new(self) -> "GATv2Model":
+        """
+        Create a new instance of the model with the same parameters.
+        This does not copy the estimator, only the configuration.
+        """
+        return self.__class__(**self.dict(exclude={"estimator"}))
```
Comment (smcolby) [inline openadmet/models/architecture/gat.py:384]: The `dict` method is deprecated; use `model_dump` instead. Deprecated in Pydantic V2.0 to be removed in V3.0.

File: openadmet/models/architecture/gat.py [hunk 2]
```diff
@@ -1,110 +1,197 @@
...
+    @classmethod
+    def validate_loss_function(cls, value):
+        """Validate loss function"""
+        allowed = ["mse", "mae", "huber", "bce", "cross_entropy"]
+        if value not in allowed:
+            raise ValueError(f"Loss function must be one of {allowed}")
+        return value
+    
+
     def __init__(
```
[the remark this answers is not in the document]
Comment (smcolby) [reply openadmet/models/architecture/gat.py:83]: Actually maybe not... our point of departure, ChemProp's `MPNN` is also a `pl.LightningModule` and makes use of an `__init__` and `super.__init__()` (the latter of which I think is necessary to link up the underlying `torch.nn.Module`...

File: openadmet/models/architecture/gat.py [hunk 3]
```diff
@@ -44,67 +111,87 @@ def __init__(
...
             # Last layer
             else:
-                in_channels = hidden_dim * num_heads if concat_heads else hidden_dim
-                out_channels = hidden_dim
+                in_channels = (
+                    self.hidden_dim * self.num_heads
+                    if self.concat_heads
+                    else self.hidden_dim
+                )
+                out_channels = self.hidden_dim
```
Comment (smcolby) [inline openadmet/models/architecture/gat.py:158]: Why call this `out_channels` when we can just directly refer to `self.hidden_dim`?

File: openadmet/models/architecture/gat.py [hunk 4]
```diff
@@ -44,67 +111,87 @@ def __init__(
...
 
         # Pooling function
-        if pooling == "mean":
-            self.pool = global_mean_pool
-        elif pooling == "max":
-            self.pool = global_max_pool
-        elif pooling == "add":
-            self.pool = global_add_pool
-        else:
-            raise ValueError(f"Unsupported pooling method: {pooling}")
```
Comment (smcolby) [inline openadmet/models/architecture/gat.py:107]: For awareness @yxnyu , I added a `validate_pooling` field validator and `_POOLING` lookup dict, hence no longer needing this block

File: openadmet/models/architecture/gat.py [hunk 5]
```diff
@@ -232,40 +347,42 @@ def build(self, scaler=None, **kwargs):
...
+                optimizer=self.optimizer,
+                optimizer_lr=self.optimizer_lr,
+                optimizer_weight_decay=self.optimizer_weight_decay,
+                scheduler=self.scheduler,
+                scheduler_factor=self.scheduler_factor,
+                scheduler_patience=self.scheduler_patience,
+                monitor_metric=self.monitor_metric,
+            )
+
+            logger.info(f"Built GATv2Model with config: {self.estimator.__dict__}")
```
Comment (smcolby) [inline openadmet/models/architecture/gat.py:374]: Per next comment, should `__dict__` be replaced by `model_dump`? Need to familiarize myself with benefits of the latter over the former.
